### PR TITLE
[REF] Migrate from `setup.py` to `pyproject.toml` (fixes #122)

### LINKED ISSUE #122: Error when installing `curvlinops` as dependency
From `laplace-torch` Github actions:

```
Collecting curvlinops-for-pytorch@ git+https://github.com/f-dangel/curvlinops (from laplace-torch==0.1a2)
  Cloning https://github.com/f-dangel/curvlinops to /tmp/pip-install-c9oemcf1/curvlinops-for-pytorch_8092676930d443ea8493f3627b08db31
  Running command git clone --filter=blob:none --quiet https://github.com/f-dangel/curvlinops /tmp/pip-install-c9oemcf1/curvlinops-for-pytorch_8092676930d443ea8493f3627b08db31
  Resolved https://github.com/f-dangel/curvlinops to commit 84c2ce[75](https://github.com/aleximmer/Laplace/actions/runs/9799308281/job/27059312114?pr=202#step:4:76)5c2c5325330e115a0d46f13ef15cfd2e
  Installing build dependencies: started
  Installing build dependencies: finished with status 'done'
  Getting requirements to build wheel: started
  Getting requirements to build wheel: finished with status 'error'
  error: subprocess-exited-with-error
  
  × Getting requirements to build wheel did not run successfully.
  │ exit code: 1
  ╰─> [17 lines of output]
      Traceback (most recent call last):
        File "/opt/hostedtoolcache/Python/3.9.19/x64/lib/python3.9/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 353, in <module>
          main()
        File "/opt/hostedtoolcache/Python/3.9.19/x64/lib/python3.9/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 335, in main
          json_out['return_val'] = hook(**hook_input['kwargs'])
        File "/opt/hostedtoolcache/Python/3.9.19/x64/lib/python3.9/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 118, in get_requires_for_build_wheel
          return hook(config_settings)
        File "/tmp/pip-build-env-oh_[92](https://github.com/aleximmer/Laplace/actions/runs/9799308281/job/27059312114?pr=202#step:4:93)17v/overlay/lib/python3.9/site-packages/setuptools/build_meta.py", line 327, in get_requires_for_build_wheel
          return self._get_build_requires(config_settings, requirements=[])
        File "/tmp/pip-build-env-oh_9217v/overlay/lib/python3.9/site-packages/setuptools/build_meta.py", line 297, in _get_build_requires
          self.run_setup()
        File "/tmp/pip-build-env-oh_9217v/overlay/lib/python3.9/site-packages/setuptools/build_meta.py", line 497, in run_setup
          super().run_setup(setup_script=setup_script)
        File "/tmp/pip-build-env-oh_9217v/overlay/lib/python3.9/site-packages/setuptools/build_meta.py", line 313, in run_setup
          exec(code, locals())
        File "<string>", line 9, in <module>
      ModuleNotFoundError: No module named 'packaging'
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
error: subprocess-exited-with-error
```

### PR DESCRIPTION
**Pull request by kourbou**
Hello! I agree with @wiseodd that requiring packaging to be installed during build time makes it difficult to use curvlinops in downstream packages such as laplace-torch, and makes the installation process a little confusing. 

I've moved the version check in `setup.py` directly into `pyproject.toml` using the new standard for setuptools-based packages:

* https://packaging.python.org/en/latest/guides/writing-pyproject-toml
* https://setuptools.pypa.io/en/latest/userguide/pyproject_config.html

Unfortunately, flake8 and darglint do not yet support `pyproject.toml` but their flags can remain in the `setup.cfg` file until support is added. (Although, I recommend eventually migrating to ruff since it can accomplish both tasks.)

* https://github.com/PyCQA/flake8/issues/234
* https://github.com/terrencepreilly/darglint/issues/130

Fixes #122.

### CODE REVIEW

**Comment by wiseodd** [comment]
Thanks @kourbou for making my job easier! I think @f-dangel won't be mad if you also replace `black`, `darglint`, `flake`, etc. with `ruff` in this PR ;)

**Comment by f-dangel** [comment]
Hi, thanks for setting this up!

Please go ahead and introduce `ruff` if you like.

I'll have time to review this during the week. @wiseodd could you also take a look?

**Comment by wiseodd** [comment]
LGTM so far. One thing I'm confused about is why doesn't this have the same problem as before, i.e. requiring `packaging`. The build backend remains the same and this PR is supposedly only translating `setup.cfg` -> `pyproject.toml`.

**Comment by kourbou** [comment]
@f-dangel Okay, I can look into adding ruff into the project later this month, although I would prefer if it was a separate PR if it's okay with you since the tool is quite opinionated and it might involve discussing lint rules and the such.

@wiseodd AFAIU the requirement on packaging was solely for the purpose of checking the version of setuptools. It was previously imported inside of `setup.py` to parse the version number but I've removed `setup.py` entirely since we can now make this check directly from `pyproject.toml`. I've also bumped the version to guarantee `pyproject.toml` support.

```
[build-system]
requires = ["setuptools >= 61.0", "setuptools_scm"]
...
```

File: .readthedocs.yaml [hunk 1]
```diff
@@ -12,7 +12,7 @@ build:
     python: "3.8"
   jobs:
     post_install:
-    - pip install --upgrade pip packaging wheel setuptools==69.5.1
+    - pip install --upgrade pip wheel setuptools==69.5.1
```
Comment (f-dangel) [inline .readthedocs.yaml:15]: I think this part should not be required anymore.

**Review by f-dangel — COMMENTED** [review]
One nit, otherwise looks good to me!

**Comment by f-dangel** [comment]
> @f-dangel Okay, I can look into adding ruff into the project later this month, although I would prefer if it was a separate PR if it's okay with you since the tool is quite opinionated and it might involve discussing lint rules and the such.

Sounds good.

File: .readthedocs.yaml (hunk 1 above)
Comment (kourbou) [reply .readthedocs.yaml]: Okay, I wasn't sure since you seem to be pinning `setuptools` to v69.5.1 ...

**Comment by kourbou** [comment]
> One nit, otherwise looks good to me!

I removed the post-install step in `.readthedocs.yaml`, it seems to work without the version pinned now. Will come back with another PR for Ruff.

**Comment by f-dangel** [comment]
Thanks! I am adding 'create a new release' to my TODO list for the coming weeks.

**Comment by kourbou** [comment]
@f-dangel Unfortunately I don't have time to make a PR and test if [Ruff] can replace all the project's linting needs, but adding it as a dependency and pasting the following into `pyproject.toml` should get you 90% of the way there:

[Ruff]: https://docs.astral.sh/ruff/

```toml
[tool.ruff]
# An extremely fast Python linter, written in Rust.
# https://docs.astral.sh/ruff/

# Additional file patterns to omit from formatting and linting.
extend-exclude = ["docs/rtd"]

# Line length to use when enforcing long-lines violations.
line-length = 97 # = 88 + 10% (B950)

[tool.ruff.lint]

# Rule codes or prefixes to enable, in addition to those specified by `select`.
extend-select = [
    "B",   # flake8-bugbear
    "C",   # pylint-convention
    "E",   # pycodestyle-error
    "W",   # pycodestyle-warning
    "C90", # mccabe
    "I",   # isort
    "DOC", # pydocstyle
    # Not implemented: https://github.com/astral-sh/ruff/issues/5713
    # "P",   # flake8-string-format
    # Not implemented: https://github.com/astral-sh/ruff/issues/458
    # "DAR", # darglint
]

# Rule codes or prefixes to ignore.
ignore = [
    "C408", # unnecessary-collection-call
    "E203", # whitespace-before-punctuation
    "E231", # missing-whitespace
    "W291", # trailing-whitespace
    "B905", # zip-without-explicit-strict
]

# Mappings from file pattern to rule codes or prefixes to exclude.
per-file-ignores = { "test/**.py" = ["I001"] }

[tool.ruff.lint.mccabe]

# The maximum McCabe complexity to allow before triggering C901 errors.
max-complexity = 10

[tool.ruff.lint.pydocstyle]

# Whether to use Google-style, NumPy-style conventions, or the PEP 257 defaults
# when analyzing docstring sections.
convention = "google"

```

**Pull request merged**